fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/properties/RecordLayerPropertyStorage.java

```diff
@@ -0,0 +1,105 @@
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.properties;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.google.common.collect.ImmutableMap;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Adopter uses this storage of mapping of {@link RecordLayerPropertyKey} to {@link RecordLayerPropertyValue} to configure parameters for Record Layer.
+ * The instances of {@link RecordLayerPropertyKey} and how they are used are defined in Record Layer.
+ * Adopters can populate {@link RecordLayerPropertyValue} and put them into this storage.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class RecordLayerPropertyStorage {
+    private static final RecordLayerPropertyStorage EMPTY_PROPERTY_STORAGE = new RecordLayerPropertyStorage(ImmutableMap.of());
+
+    @Nonnull
+    private final ImmutableMap<RecordLayerPropertyKey<?>, RecordLayerPropertyValue<?>> propertyMap;
+
+    private RecordLayerPropertyStorage(@Nonnull ImmutableMap<RecordLayerPropertyKey<?>, RecordLayerPropertyValue<?>> propertyMap) {
+        this.propertyMap = propertyMap;
+    }
+
+    @Nonnull
+    public ImmutableMap<RecordLayerPropertyKey<?>, RecordLayerPropertyValue<?>> getPropertyMap() {
+        return propertyMap;
+    }
+
+    @Nonnull
+    public <T> T getPropertyValue(@Nonnull RecordLayerPropertyKey<T> propertyKey) {
+        if (propertyMap.containsKey(propertyKey)) {
+            Object value = propertyMap.get(propertyKey).getValue();
+            try {
+                return propertyKey.getType().cast(value);
+            } catch (ClassCastException ex) {
+                throw new RecordCoreException("Invalid type for record context property", ex)
+                        .addLogInfo(LogMessageKeys.PROPERTY_NAME, propertyKey.getName(),
+                                LogMessageKeys.PROPERTY_TYPE, value.getClass().getName());
+            }
+        } else {
+            return propertyKey.getDefaultValue();
+        }
+    }
+
+    @Nonnull
+    public Builder toBuilder() {
+        return new Builder(propertyMap);
+    }
+
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    public static RecordLayerPropertyStorage getEmptyInstance() {
+        return EMPTY_PROPERTY_STORAGE;
+    }
+
+    public static class Builder {
+        private final Map<RecordLayerPropertyKey<?>, RecordLayerPropertyValue<?>> propertyMap;
+
+        private Builder() {
+            this.propertyMap = new HashMap<>();
+        }
+
+        private Builder(ImmutableMap<RecordLayerPropertyKey<?>, RecordLayerPropertyValue<?>> properties) {
+            this.propertyMap = new HashMap<>(properties);
+        }
+
+        public Builder addProp(@Nonnull RecordLayerPropertyValue<?> propValue) {
```

Comment:
This might not be necessary, but it might be nice to provide helper methods:

```java
   public <T> Builder addProp(@Nonnull RecordLayerPropertyKey<T> key, @Nullable T value);
   public <T> Builder addProp(@Nonnull RecordLayerPropertyKey<T> key, @Nullable Supplier<T> valueSupplier);
```

Each of those would construct the appropriate `RecordLayerPropertyValue` based on the provided key. But these are essentially just convenience methods that are essentially just syntactic sugar around existing functionally.

We could also leave this off (at least for now) if we want to decrease the surface area of this new change

Reply:
Added these 2 helper methods.